### PR TITLE
[LOOP-26] Upload loop dosing decision data to Tidepool backend

### PR DESCRIPTION
**Pull request by darinkrauss**
- https://tidepool.atlassian.net/browse/LOOP-26
- Make bolusUnits optional in AutomaticDoseRecommendation
- Unify DosingDecisionStore in LoopKit from Loop
- Remove unnecessary Codable support for various Errors
- Return StoredGlucoseSamples after adding via DeviceDataManager
- Update StoredDosingDecision properties
- Refactor how errors and warnings are captured in StoredDosingDecision
- Add more explicit LoopError cases
- Add LoopWarning
- Use explicit LoopError or PumpManagerError where possible
- Add reason for various dosing decisions
- Explicit creation of dosing decision when cancelling temp basal
- Update LoopDataManager.update to return error rather than throw
- Update LoopDataManager.updatePredictedGlucoseAndRecommendedDose to return error rather than throw
- Inject updatable dosing decision into LoopDataManager loop
- Capture all errors during LoopDataManager loop rather than just first
- Update tests

Note: This one is going to require approval from @ps2 at a minimum, but would appreciate any other feedback.

### CODE REVIEW

File: Loop/Managers/LoopDataManager.swift [hunk 1]
```diff
@@ -758,57 +779,26 @@ extension LoopDataManager {
...
-                                                                  deviceSettings: UIDevice.current.deviceSettings,
-                                                                  errors: [error, state.error, insulinOnBoardError].compactMap { $0 })
-                        self.dosingDecisionStore.storeDosingDecision(dosingDecision) {}
-                    }
-                }
-            }
-        }
-    }
-
     func storeBolusDosingDecision(_ bolusDosingDecision: BolusDosingDecision, withDate date: Date) {
```
Comment (ps2) [inline Loop/Managers/LoopDataManager.swift]: Would a better name for this be `storeManualBolusDosingDecision`?

File: Loop/Managers/LoopDataManager.swift [hunk 2]
```diff
@@ -862,60 +852,86 @@ extension LoopDataManager {
...
+        logger.default("Loop ended")
+        notify(forChange: .tempBasal)
+    }
+
+    fileprivate enum UpdateReason: String {
+        case loop
+        case getLoopState
     }
 
     /// - Throws:
```
Comment (ps2) [inline Loop/Managers/LoopDataManager.swift]: Remove this documentation, now that it's no longer throwing?

**Review by ps2 — APPROVED** [review]
There's a lot here, and some pretty critical areas. It looks ok to me; just a few nits. I think this code deserves some thorough QA testing around general loop decision making, just because so much logic changed, even if it was mostly refactoring.

**Review by rickpasetto — APPROVED** [review]
LGTM

**Pull request merged**